File: C++/프로그래머스/2/84512. 모음 사전/모음 사전.cpp
#include <string>
#include <vector>
#include <iostream>

using namespace std;

int solution(string word) {
    vector<char> vowels = {'A', 'E', 'I', 'O', 'U'};
    // 5^4 = 625, 5^3 = 125, 5^2 = 25, 5^1 = 5, 5^0 = 1
    vector<int> power = {781, 156, 31, 6, 1};  // 각 자리수에서 가능한 단어의 수

    int result = 0;
    
    // 주어진 단어의 각 문자의 위치를 계산
    for (int i = 0; i < word.size(); i++) {
        // 현재 문자가 vowels 벡터에서 몇 번째에 위치하는지 찾음
        for (int j = 0; j < 5; j++) {
            if (word[i] == vowels[j]) {
                // 위치에 따른 사전 순서를 계산
                result += power[i] * j + 1;
                break;
            }
        }
    }
    
    return result;
}

File: Python3/프로그래머스/2/84512. 모음 사전/모음 사전.py
import itertools
# permutations : 순열
# prodcut : 중복 순열
# combinations
# combinations_with_replacement # 중복 조합

def solution(word):
    arr = [] # 만들 수 있는 모든 사전 조합
    words = ['A', 'E', 'I', 'O', 'U'] # 사용 가능한 문자
    
    # 1글자부터 5글자까지의 중복 순열 생성
    for j in range(1, 6):
        for i in itertools.product(words, repeat = j):
            # 생성된 조합을 리스트에 추가
            arr.append(list(i))
    
    # 사전 순서로 정렬
    arr.sort()

    answer = 0
    # 모든 조합을 확인하며 단어를 찾기
    for i in arr:
        answer += 1
        # 리스트를 문자열로 변환하여 비교
        st = ''
        for s in i:
            st += s
        if st == word:
            break
    
    return answer

